hal/halx86/apic/apicsmp.c

```diff
@@ -138,3 +130,172 @@ ApicStartApplicationProcessor(
+                               APIC_MT_Fixed,
+                               APIC_TGM_Edge,
+                               DestinationShortHand);
+}
+
+/*!
+ *  \brief Requests an IPI with a specified vector on the specified processors.
+ *
+ *  \param TargetSet - Specifies the set of processors to send the IPI to.
+ *  \param Vector - Specifies the interrupt vector to be delivered.
+ *
+ *  \remarks This function is exported on Windows 10.
+ */
+VOID
+NTAPI
+HalRequestIpiSpecifyVector(
+    _In_ KAFFINITY TargetSet,
+    _In_ UCHAR Vector)
+{
+    KAFFINITY ActiveProcessors = KeQueryActiveProcessors();
+    KAFFINITY RemainingSet, SetMember;
+    ULONG ProcessorIndex;
+    ULONG LApicId;
+
+    /* Sanitize the target set */
+    TargetSet &= ActiveProcessors;
+
+    /* Check if all processors are requested */
+    if (TargetSet == ActiveProcessors)
+    {
+        /* Send an IPI to all processors, including this processor */
+        HalpBroadcastIpiSpecifyVector(Vector, TRUE);
+        return;
+    }
+
+    /* Check if all processors except the current one are requested */
+    if (TargetSet == (ActiveProcessors & ~KeGetCurrentPrcb()->SetMember))
+    {
+        /* Send an IPI to all processors, excluding this processor */
+        HalpBroadcastIpiSpecifyVector(Vector, FALSE);
+        return;
+    }
+
+    /* Loop while we have more processors */
+    RemainingSet = TargetSet;
+    while (RemainingSet != 0)
+    {
+        NT_VERIFY(BitScanForwardAffinity(&ProcessorIndex, RemainingSet) != 0);
+        ASSERT(ProcessorIndex < KeNumberProcessors);
+        SetMember = AFFINITY_MASK(ProcessorIndex);
+        RemainingSet &= ~SetMember;
+
+        /* Send the interrupt to the target processor */
+        LApicId = HalpProcessorIdentity[ProcessorIndex].LapicId;
+        ApicRequestGlobalInterrupt(LApicId,
+                                   Vector,
+                                   APIC_MT_Fixed,
+                                   APIC_TGM_Edge,
+                                   APIC_DSH_Destination);
+    }
+}
+
+/*!
+ *  \brief Requests an IPI interrupt on the specified processors.
+ *
+ *  \param TargetSet - Specifies the set of processors to send the IPI to.
+ */
+VOID
+NTAPI
+HalpRequestIpi(
+    _In_ KAFFINITY TargetSet)
+{
+    /* Request the IPI vector */
+    HalRequestIpiSpecifyVector(TargetSet, APIC_IPI_VECTOR);
+}
+
+#ifdef _M_AMD64
+
+/*!
+ *  \brief Requests a software interrupt on the specified processors.
+ *
+ *  \param TargetSet - Specifies the set of processors to send the IPI to.
+ *  \param Irql - Specifies the IRQL of the software interrupt.
+ */
+VOID
+NTAPI
+HalpSendSoftwareInterrupt(
+    _In_ KAFFINITY TargetSet,
+    _In_ KIRQL Irql)
+{
+    UCHAR Vector;
+
+    /* Get the vector for the requested IRQL */
+    if (Irql == APC_LEVEL)
+    {
+        Vector = APC_VECTOR;
+    }
+    else if (Irql == DISPATCH_LEVEL)
+    {
+        Vector = DISPATCH_VECTOR;
+    }
+    else
+    {
+        ASSERT(FALSE);
```

Comment:
It's not possible to have other Irqls given to that function?

Reply:
No, that are the only 2 software interrupts.